### PR TITLE
Interactive init

### PR DESCRIPTION
**Pull request by phamann**
### TL;DR
Refactors `compute init` to accept its input interactively and act as a setup wizard to provision a service, backend and domain.

### Why?
We have observed friction in user testing and feedback that a users has to create a service then attach a domain and backend before a compute package can be deployed and activated. If you are new to Fastly this may be confusing. Therefore, we think that provisioning all of this for you within a single command will reduce this on-boarding friction.

### What?
- Refactor command to take input interactively
- Refactor `fastly configure` to persist user email to global config file so that it can be read by `compute init` for default author email.
- Add a `NullProgress` writer which swallows all writes
- Add an `UndoStack` which we use to teardown any state we have created in the API if an error occurs.
- Add [`golang-petname`](https://github.com/dustinkirkland/golang-petname) dep to randomly generate nice domain names

### Example:
Screenshot 2020-03-09 at 12 23 28

### CODE REVIEW

File: pkg/common/file.go [hunk 1]
```diff
@@ -55,7 +55,7 @@ func CopyFile(src, dst string) (err error) {
 	if err != nil {
 		return fmt.Errorf("error reading source file: %w", err)
 	}
-	defer in.Close()
+	defer in.Close() // #nosec G307
```
Comment (peterbourgon) [inline pkg/common/file.go]: 🙄 IMO disable this check entirely. (Minor nit, not blocking this PR.)

File: pkg/common/undo.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+// Push method pushes an Undoer element onto the UndoStack.
+func (s *UndoStack) Push(elem UndoFn) {
+	s.states = append(s.states, elem)
+}
+
+// Len method returns the number of elements in the UndoStack.
+func (s *UndoStack) Len() int {
+	return len(s.states)
+}
```
Comment (peterbourgon) [inline pkg/common/undo.go]: If you are going to go through the (appreciated!) trouble of defining an UndoStack type, IMO you should go all the way and also provide a helper method to execute it in case of error.

```go
func (s *UndoStack) RunIfError(err error) {
	if err == nil {
		return
	}
	for i := len(s.states)-1; i >= 0; i-- {
		s.states[i]()
	}
}
```

That means you could reduce the public API to merely `Push` and `RunIfError`, probably remove the UndoFn type and use plain `func()`, and use it in-situ like

```go
func f() (err error) {
	var undo common.UndoStack
	defer undo.RunIfError(err)
	
	undo.Push(func() { ... })
```

File: pkg/compute/init.go [hunk 3]
```diff
@@ -4,32 +4,59 @@ import (
...
+
+const (
+	defaultTemplate       = "https://github.com/fastly/fastly-template-rust-default.git"
+	defaultTopLevelDomain = "edgecompute.app"
+	manageServiceBaseURL  = "https://manage.fastly.com/configure/services/"
+)
 
 var (
+	gitRepositoryRegEx        = regexp.MustCompile(`((git|ssh|http(s)?)|(git@[\w\.]+))(:(//)?)([\w\.@\:/\-~]+)(\.git)(/)?`)
+	domainNameRegEx           = regexp.MustCompile(`(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\.)+[a-z0-9][a-z0-9-]{0,61}[a-z0-9]`)
```
Comment (peterbourgon) [inline pkg/compute/init.go]: I am afraid this is incomplete, but I don't have the strength/knowledge/intelligence to test it. Does it come from some known source? Is it possible to be abused?

File: pkg/compute/init.go [hunk 4]
```diff
@@ -75,19 +127,143 @@ func (c *InitCommand) Exec(in io.Reader, out io.Writer) (err error) {
...
+		}
+		if c.author == "" {
+			c.author = defaultEmail
+		}
+	}
+
+	if c.from == "" {
+		text.Output(out, "%s", text.Bold("Template:"))
+		for i, template := range defaultTemplates {
+			text.Output(out, "[%d] %s", i, template.Name)
```
Comment (peterbourgon) [inline pkg/compute/init.go]: IMO it would be nice to put the URL of the template in parenthesis here, so that the " . . . or type URL" portion of the prompt is visually reinforced.

File: pkg/compute/init.go [hunk 5]
```diff
@@ -75,19 +127,143 @@ func (c *InitCommand) Exec(in io.Reader, out io.Writer) (err error) {
...
+	_, err = c.Globals.Client.CreateBackend(&fastly.CreateBackendInput{
+		Service: service.ID,
+		Version: 1,
+		Name:    c.backend,
+		Address: c.backend,
+	})
+	if err != nil {
+		return fmt.Errorf("error creating backend: %w", err)
+	}
+	undoStack.Push(func() error {
```
Comment (peterbourgon) [inline pkg/compute/init.go]: If you want undo functions to return an error — which is sensible — then logging each non-nil error when executing the undo stack would probably be the best bet.

File: pkg/compute/init.go [hunk 6]
```diff
@@ -134,22 +310,39 @@ func (c *InitCommand) Exec(in io.Reader, out io.Writer) (err error) {
...
 
 	progress.Done()
-	text.Success(out, "Initialized package %s to %s", m.Name, abspath)
+
+	text.Break(out)
+
+	fmt.Fprintf(out, "Initialized package %s to:\n\t%s\n\n", text.Bold(m.Name), text.Bold(abspath))
+	fmt.Fprintf(out, "Manage this service at:\n\t%s\n\n", text.Bold(fmt.Sprintf("%s%s", manageServiceBaseURL, service.ID)))
+	fmt.Fprintf(out, "To compile the package, run:\n\t%s\n\n", text.Bold("fastly compute build"))
+	fmt.Fprintf(out, "To deploy the package, run:\n\t%s\n\n", text.Bold("fastly compute deploy"))
```
Comment (peterbourgon) [inline pkg/compute/init.go]: Is the idiom of emitting "output" from a larger process by doing one or more

> Prompt:
>&nbsp;&nbsp;&nbsp;&nbsp;**Output**

blocks at the conclusion of the process done elsewhere in the CLI? If so, is there an opportunity to make this style consistent?

File: pkg/text/progress.go [hunk 7]
```diff
@@ -234,3 +234,33 @@ func (p *VerboseProgress) Done() {}
...
 // Fail implements the Progress interface. It's a no-op.
 func (p *VerboseProgress) Fail() {}
+
+//
+//
+//
+
+// NullProgress is an implementation of Progress which discards everything
+// written into it and produces no output.
+type NullProgress struct{}
```
Comment (peterbourgon) [inline pkg/text/progress.go]: 🤘

File: pkg/common/undo.go (hunk 2 above)
Comment (peterbourgon) [reply pkg/common/undo.go]: If the undo funcs returned an error, RunIfError could take an io.Writer to which error results could be printed.

Comment (phamann) [reply pkg/common/undo.go]: Added this helper method in https://github.com/fastly/cli/pull/5/commits/d4b73e21737f74615747e148b2902b1ccaca7ff2

File: pkg/compute/init.go (hunk 3 above)
Comment (phamann) [reply pkg/compute/init.go]: It seemed the most robust in the top SO response for this question, however I agree with you that it could be brittle and I'm sad the Go standard library doesn't offer something such as how you can parse and validate IP addresses.

File: pkg/compute/init.go (hunk 5 above)
Comment (phamann) [reply pkg/compute/init.go]: Added error logging as part of d4b73e2

File: pkg/compute/init.go (hunk 6 above)
Comment (phamann) [reply pkg/compute/init.go]: Added a `text.Description()` helper in https://github.com/fastly/cli/pull/5/commits/97bdd8200a757642fb81ed9ee0ca1a8ece6a0b2a

File: pkg/compute/init.go (hunk 4 above)
Comment (phamann) [reply pkg/compute/init.go]: Added in https://github.com/fastly/cli/pull/5/commits/78fd8c262bf2a19bd64543fc003bdbbff37d04b7

File: pkg/common/file.go (hunk 1 above)
Comment (phamann) [reply pkg/common/file.go]: Will do in another PR and discuss there.
